docs/design/storage-auto-scale.md

```diff
@@ -0,0 +1,83 @@
+        namespace: openshift-storage
+    spec:
+        storageCluster: 
+            name: <storageclusterName>
+        deviceClassName: ssd
+        clusterCapacityLimit: <capacity>
+        maxOsdSize: 8Ti
+        timeoutSeconds: 1800   
+    status:
+        phase: (NotStarted | InProgress | Failed | Succeeded)
+        LastRunTimestamp: (only on Failed | Succeeded)
+        currentOsdSize: 2TiB
+        expectedOsdSize: 4TiB
+        currentOsdCount: 3​
+        expectedOsdCount: 3
+    ```
+
+- Watch prometheus periodically every 10 mins and depending on the Osd fill percentage, reconcile the StorageAutoScaling CR.
+  
+- If the Osd size is less than maxOsdSize(default:8Tib), do vertical scaling by doubling the each osd sizes for that device class.
+
+- If the Osd size is equal to maxOsdSize(default:8Tib), do a horizontal scaling, by adding 1 osd of 8Tib on each storage node, by simply increasing the (count+1) in the device set associated with a device class.   
+
+## Low-Level Design 
+
+1) Create a controller that requeueAfter(https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/reconcile#Result) every 10 mins.
+   
+2) List all the device class Osds used percentage using Prometheus metrics.
+    Query: (ceph_osd_metadata * on (ceph_daemon, namespace, managedBy) group_right(device_class,hostname) (ceph_osd_stat_bytes_used / ceph_osd_stat_bytes))
+       `OsdPercentage`:
+        ```
+        osd1-ssd               69%
+        osd2-ssd               72%
+        ```
+
+3) If the highest osd percentage reported is more than 70%(reaching osd nearfull), scaling is needed.
+
+4) If scaling is needed calculate the `expectedOsdSize` and `expectedOsdCount`.
+    1)  Calculate the total number osds and osd size from the prometheus response.
+    2)  Calculate the total number osds and osd size from the storageCluster.
+    3)  If the above values are equal(needed to avoid double re-sizing), create or update the `OsdPercentage map`
+        1)  If the Osd size is less than maxOsdSize(default:8Tib), do vertical scaling by doubling the each osd sizes for that device class.
+        2)  If the Osd size is equal to maxOsdSize(default:8Tib), do a horizontal scaling, by adding 1 osd of maxOsdSize(default:8Tib) on each storage node.
+      
+            `OsdPercentage` map:
+
+            ```
+                expectedOsdSize: 4
+                expectedOsdCount: 6
+            ```
+
+5) Depending on the `OsdPercentage map` set the `expectedOsdSize` and `expectedOsdCount` and on the StorageAutoScaling status to reflect the new expected value. And update `phase` to `InProgress` on the `StorageAutoScaling` CR which need scaling.
+
+6) Scale with the following algorithm:
+   1) Check if the totalCapacity allows to scale it or not.
+   2) Patch the Storagecluster, with all the device sets update needed at the same time.
+   3) Verify the Storagecluster whether the new osds are added or scaled in size, for all the device sets. 
```

Comment:
Seems like no need for verification, if the patch in step 2 successfully updated?

Reply:
we need verification wether the scaling is done successfully or not, as the actual scaling is dependent on 3rd party apis

Reply:
Ok, I see you're just referring to checking for the completion so we can update the status. Let's just be clear that checking the status happens after we requeue, rather than waiting for completion here.

Reply:
as discussed offline, we need this as we need to update the startOSD size and count in the stauts